# Project 1 -- Niharika Rai

**TA Help:** (for instance) John Smith, Alice Jones, etc., list names of any TAs who helped you

- For example: Help with figuring out how to write a function (describe the tasks that they helped you with)

**Collaboration:** My Friend in CS, My Uncle, Another Student, etc., list names of any other people who helped you

(describe the tasks that they helped you with)
- For example: helped figuring out how to load the dataset.
- Another example: helped debug error with my plot.

**Internet Resources:** Stack Exchange, Stack Overflow, etc.

(describe any information that you learned from internet resources, including the URLs)
- data frames in Pandas versus R from StackOverflow  https://stackoverflow.com/questions/8991709/why-were-pandas-merges-in-python-faster-than-data-table-merges-in-r-in-2012

**ChatGPT, Gemini, Claude, etc:** Any language models or generative AI chatbots that helped you.

(if you used any such tools, please tell us here)
- For example:  I asked ChatGPT how to define a new data frames
- Another example:  Gemini told me how to make a function for sorting my data

**OVERALL MESSAGE:** Any time that you used anything except your brain to solve the questions in these projects, you need to disclose such resources at the start of the project, with details about your usage of the tools.

**YOUR OWN WORK:** Even when you utilize other resources, do NOT just copy and paste.  Write all explanations in your own words, using several sentences in English, which are understandable and which you wrote (and did not just copy and paste).

## Question 1

In [6]:
%%bash
echo Hello World!  

Hello World!


In [15]:
%%bash
head -n 1 /anvil/projects/tdm/data/bay_area_bike_share/kaggle/status.csv | awk -F',' '{print NF}'

4


In [20]:
ls -lh /anvil/projects/tdm/data/bay_area_bike_share/kaggle/status.csv

-rw-r--r-- 1 nobody nogroup 1.9G Nov 14  2019 /anvil/projects/tdm/data/bay_area_bike_share/kaggle/status.csv


The dataset has 4 columns, and the file size is 1.9 GB. To find answers to these questions, I asked ChatGPT on how to load the dataset and find the command for displaying the file size using bash (I have never used or heard of it). To load the dataset, I understood that the 'head -n 1' will print the first line, which consists of its column names. The 'awk -F',' '{print NF}'' command will treat commas as field separators, once the dataset has been loaded, and will count the number of columns. And lastly, the 'ls -lh' command will display basic information about the dataset such as its owner, group, size, date last modified, etc. From my output, I can tell that the file size is 1.9 gigabytes. 

## Question 2

In [6]:
%%bash 
cd /anvil/projects/tdm/data/bay_area_bike_share/baywheels 
ls
pwd

2017-fordgobike-tripdata.csv
201801-fordgobike-tripdata.csv
201802-fordgobike-tripdata.csv
201803-fordgobike-tripdata.csv
201804-fordgobike-tripdata.csv
201805-fordgobike-tripdata.csv
201806-fordgobike-tripdata.csv
201807-fordgobike-tripdata.csv
201808-fordgobike-tripdata.csv
201809-fordgobike-tripdata.csv
201810-fordgobike-tripdata.csv
201811-fordgobike-tripdata.csv
201812-fordgobike-tripdata.csv
201901-fordgobike-tripdata.csv
201902-fordgobike-tripdata.csv
201903-fordgobike-tripdata.csv
201904-fordgobike-tripdata.csv
201905-baywheels-tripdata.csv
201906-baywheels-tripdata.csv
201907-baywheels-tripdata.csv
201908-baywheels-tripdata.csv
201909-baywheels-tripdata.csv
201910-baywheels-tripdata.csv
201911-baywheels-tripdata.csv
201912-baywheels-tripdata.csv
202001-baywheels-tripdata.csv
202002-baywheels-tripdata.csv
202003-baywheels-tripdata.csv
202005-baywheels-tripdata.csv
202006-baywheels-tripdata.csv
202007-baywheels-tripdata.csv
202008-baywheels-tripdata.csv
202009-baywheels-tripdata

In [7]:
%%bash
cd /anvil/projects/tdm/data/election
ls
pwd

escaped2020sample.txt
itcont1980.txt
itcont1982.txt
itcont1984.txt
itcont1986.txt
itcont1988.txt
itcont1990.txt
itcont1992.txt
itcont1994.txt
itcont1996.txt
itcont1998.txt
itcont2000.txt
itcont2002.txt
itcont2004.txt
itcont2006.txt
itcont2008.txt
itcont2010.txt
itcont2012.txt
itcont2014.txt
itcont2016.txt
itcont2018.txt
itcont2020.txt
itcont2022.txt
itcont2024.txt
/anvil/projects/tdm/data/election


I watched Dr. Ward's video for this question and used the 'cd', 'ls -la', and 'pwd' commands. However, the answer to the months and years for the files in both directories came from using the 'ls' code snippet, which displays the file names. The file names in this example contain the individual month and year. To answer the question, the Baywheels directory includes all 12 months of the year. The years range from 2017 to 2025. For the elections directory, the years range from 1980 to 2024, increasing in increments of 2. See outputs above. 

## Question 3

In [32]:
%%bash
head -n6 /anvil/projects/tdm/data/bay_area_bike_share/kaggle/trip.csv

id,duration,start_date,start_station_name,start_station_id,end_date,end_station_name,end_station_id,bike_id,subscription_type,zip_code
4576,63,8/29/2013 14:13,South Van Ness at Market,66,8/29/2013 14:14,South Van Ness at Market,66,520,Subscriber,94127
4607,70,8/29/2013 14:42,San Jose City Hall,10,8/29/2013 14:43,San Jose City Hall,10,661,Subscriber,95138
4130,71,8/29/2013 10:16,Mountain View City Hall,27,8/29/2013 10:17,Mountain View City Hall,27,48,Subscriber,97214
4251,77,8/29/2013 11:29,San Jose City Hall,10,8/29/2013 11:30,San Jose City Hall,10,26,Subscriber,95060
4299,83,8/29/2013 12:02,South Van Ness at Market,66,8/29/2013 12:04,Market at 10th,67,319,Subscriber,94103


In [33]:
%%bash 
cut -d, -f4,7 /anvil/projects/tdm/data/bay_area_bike_share/kaggle/trip.csv > $HOME/startandendlocations.csv

In [34]:
%%bash 
head $HOME/startandendlocations.csv

start_station_name,end_station_name
South Van Ness at Market,South Van Ness at Market
San Jose City Hall,San Jose City Hall
Mountain View City Hall,Mountain View City Hall
San Jose City Hall,San Jose City Hall
South Van Ness at Market,Market at 10th
Golden Gate at Polk,Golden Gate at Polk
Santa Clara at Almaden,Adobe on Almaden
San Salvador at 1st,San Salvador at 1st
South Van Ness at Market,South Van Ness at Market


I followed all the instructions presented by Dr. Ward's video for this question. I located the column numbers starting from 1. Hence, the start station name is the 4th column, and the end station name is the 7th column. I located these columns, used the cut command, -d, -f flags, and saved them in a new CSV file in my home directory. I displayed the first few columns of the new CSV file using the head command. 

## Question 4

In [42]:
%%bash 
grep "Van Ness" /anvil/projects/tdm/data/bay_area_bike_share/kaggle/trip.csv > $HOME/vanness.csv

In [43]:
%%bash 
head $HOME/vanness.csv 

4576,63,8/29/2013 14:13,South Van Ness at Market,66,8/29/2013 14:14,South Van Ness at Market,66,520,Subscriber,94127
4299,83,8/29/2013 12:02,South Van Ness at Market,66,8/29/2013 12:04,Market at 10th,67,319,Subscriber,94103
4760,113,8/29/2013 17:01,South Van Ness at Market,66,8/29/2013 17:03,South Van Ness at Market,66,553,Subscriber,94103
5070,168,8/29/2013 21:43,South Van Ness at Market,66,8/29/2013 21:46,South Van Ness at Market,66,598,Subscriber,94115
4625,175,8/29/2013 15:02,San Francisco City Hall,58,8/29/2013 15:05,South Van Ness at Market,66,267,Subscriber,94115
4721,184,8/29/2013 16:27,Market at 10th,67,8/29/2013 16:30,South Van Ness at Market,66,416,Subscriber,94107
4720,188,8/29/2013 16:27,Market at 10th,67,8/29/2013 16:30,South Van Ness at Market,66,379,Subscriber,94109
4824,207,8/29/2013 17:38,Market at 10th,67,8/29/2013 17:42,South Van Ness at Market,66,632,Subscriber,94103
4765,228,8/29/2013 17:05,South Van Ness at Market,66,8/29/2013 17:08,Market at 10th,67,553,Subscrib

I followed all the instructions presented by Dr. Ward's video for this question. I used the 'grep' command and instead put "Van Ness" which will filter out the lines that contain this phrase. I then saved it in a new file called, 'vanness.csv' in my home directory and displayed its first few rows using the head command. They all do show "Van Ness" in them at some point. 

## Question 5

In [1]:
%%bash 
head /anvil/projects/tdm/data/bay_area_bike_share/kaggle/status.csv 

station_id,bikes_available,docks_available,time
2,2,25,2013/08/29 12:06:01
2,2,25,2013/08/29 12:07:01
2,2,25,2013/08/29 12:08:01
2,2,25,2013/08/29 12:09:01
2,2,25,2013/08/29 12:10:01
2,2,25,2013/08/29 12:11:01
2,2,25,2013/08/29 12:12:01
2,2,25,2013/08/29 12:13:01
2,2,25,2013/08/29 12:15:01


In [2]:
%%bash
awk -F',' '$1 == 2 {count++} END {print count}' /anvil/projects/tdm/data/bay_area_bike_share/kaggle/status.csv

1046898


In [3]:
%%bash
awk -F',' '$1 == 3 {count++} END {print count}' /anvil/projects/tdm/data/bay_area_bike_share/kaggle/status.csv

1047113


In [4]:
%%bash
awk -F',' '$1 == 4 {count++} END {print count}' /anvil/projects/tdm/data/bay_area_bike_share/kaggle/status.csv

1047100


In [5]:
%%bash
awk -F',' '$1 == 5 {count++} END {print count}' /anvil/projects/tdm/data/bay_area_bike_share/kaggle/status.csv

1047142


I asked ChatGPT to show me how to set Boolean conditions using bash and count the matching lines. I first loaded the dataset and used the head command to display its first few rows. This helped me see the different column names and locate the position of the station_id column in the dataset. I then learnt that: 'awk -F','' sets the comma as a field separator. This helps to locate the different columns when the condition is set. Then, '$1' stands for the first column in the dataset, which is the station_id. Throughout the different lines of code, this is set equal to 2, 3, 4, and 5. The 'count++' counts the matching rows as the condition is met. The 'print count' displays the total number of matching lines when the station_id is 2, 3, 4, and 5. 

Based on the output, the matching lines are: 
station 2: 1046898 
station 3: 1047113
station 4: 1047100
station 5: 1047142 

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
